### PR TITLE
Census: seed 2019-2020 AP CS offerings

### PR DESCRIPTION
**Pull request by hacodeorg**
[PLC-821](https://codedotorg.atlassian.net/browse/PLC-821): Ingesting 2019-2020 CollegeBoard CSA and CSP data from [here](https://drive.google.com/drive/u/0/folders/16_zeAcCHNC5OIOkH40Lsgb6nBfDU4YHv).

Before we can ingest AP CS offerings, we have to first extract and ingest CollegeBoard school code-NCES code mappings.

## How to ingest College board school code-NCES code mappings
1. Copy CSV file with mappings to `ap_school_codes/<school_year>-<school_year + 1>.test` in [cdo-census](https://s3.console.aws.amazon.com/s3/buckets/cdo-census?region=us-east-1&tab=overview) S3 bucket.

2. From a Rails console in `production-console` machine, test ingesting the new file by running `Census::ApSchoolCode.dry_seed_s3_object(<school_year>, <file_extension>)`.
    * Example: Census::ApSchoolCode.dry_seed_s3_object(2019, 'test')
    * Have to test on `production-console` because the seeding process relies on matching school IDs to valid records in `School` model.

3. Change the file extension in S3 from .test to .csv. The file will be automatically ingested to staging/test/production instances the next time they rebuid. Or you can manually trigger ingestion by running `Census::ApSchoolCode.seed_from_s3`.

## How to ingest an AP CS offering file
1. Copy the file to [cdo-census](https://s3.console.aws.amazon.com/s3/buckets/cdo-census/?region=us-east-1#)  S3 bucket following this path `ap_cs_offerings/{CSA|CSP}-{year}-{year+1}.test`.
    * Example: CSP-2019-2020.test
    * Use`.test` extension or any extensions other than `.csv` so the automated seeding process won't pick up the new file.

2. From a Rails console in production-console machine, test ingesting the new file by running `Census::ApCsOffering.dry_seed_s3_object(<CSP|CSA>, <school_year>, <file_extension>)`.
    * Example: Census::ApCsOffering.dry_seed_s3_object('CSP', 2019, 'test')
    * Have to test on `production-console` because the seeding process relies on matching school codes (College board AI code) in CSV file to valid records in `Census::ApSchoolCode` table.

3. Fix any issues that occur and deploy code to production if there are code changes.

4. Change the file extension in S3 from .test to .csv. The file will be automatically ingested to staging/test/production instances the next time they rebuid. Or you can manually trigger ingestion by running `Census::ApCsOffering.seed_from_s3`.

## Links
Related PR: Ingesting state CS offerings https://github.com/code-dot-org/code-dot-org/pull/34925/

## Testing story
From a Rails console in production-console, dry-test ingesting individual files:
```
Census::ApSchoolCode.dry_seed_s3_object(2017)
Census::ApSchoolCode.dry_seed_s3_object(2019, 'test')

Census::ApCsOffering.dry_seed_s3_object('CSA', 2016)
Census::ApCsOffering.dry_seed_s3_object('CSP', 2017)
Census::ApCsOffering.dry_seed_s3_object('CSA', 2019, 'test')
Census::ApCsOffering.dry_seed_s3_object('CSP', 2019, 'test')
```

## Reviewer Checklist:
- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by bencodeorg** [comment]
Main comment -- do you have another PR to import the new crosswalk between AI codes and NCES IDs (importing to [this model](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/census/ap_school_code.rb))? Or maybe no PR was necessary.

A couple documentation ideas:
1. Could you put somewhere in code the reason for the new file extension options (ie, something akin to what you have in the PR description here about not using CSVs while testing).
2. I think your google drive link doesn't actually go to where this data is stored.
3. I think? you tested this in production console b/c you needed a full version of some data (schools? ap school codes?). Maybe clarify that in PR description.

**Comment by hacodeorg** [comment]
Ohhh the crosswalk 🤦 I will take a look. 
I'll add comments for comment 1 & 3.

> Main comment -- do you have another PR to import the new crosswalk between AI codes and NCES IDs (importing to [this model](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/census/ap_school_code.rb))? Or maybe no PR was necessary.
> 
> A couple documentation ideas:
> 
>     1. Could you put somewhere in code the reason for the new file extension options (ie, something akin to what you have in the PR description here about not using CSVs while testing).
> 
>     2. I think your google drive link doesn't actually go to where this data is stored.
> 
>     3. I think? you tested this in production console b/c you needed a full version of some data (schools? ap school codes?). Maybe clarify that in PR description.

File: dashboard/app/models/census/ap_cs_offering.rb [hunk 1]
```diff
@@ -28,35 +28,45 @@ class Census::ApCsOffering < ApplicationRecord
...
             Census::ApCsOffering.find_or_create_by!(
               ap_school_code: ap_school_code,
               course: course,
               school_year: school_year
             )
-          rescue ActiveRecord::RecordNotFound
-            # We don't have mapping for every school code so skip over any that
-            # can't be found in the database.
-            CDO.log.warn "AP CS Offering seeding: skipping unknown school code #{normalized_school_code}, school name #{raw_school_name}"
+            succeeded += 1
```
Comment (bencodeorg) [inline dashboard/app/models/census/ap_cs_offering.rb]: Don't you want to know how many would succeed/skip during a dry run?

Comment (hacodeorg) [reply dashboard/app/models/census/ap_cs_offering.rb]: you caught a 🐞

**Review by bencodeorg — APPROVED** [review]

**Pull request merged**